ground-server/src/app/page.tsx

```diff
@@ -17,31 +17,45 @@ const ResponsiveReactGridLayout = WidthProvider(Responsive);
 export default function Home() {
   const [widgets, setWidgets] = useState<Widget[]>([]);
   const [data, setData] = useState<Data[]>([]);
+  const [ws, setWs] = useState<WebSocket | null>(null); // Store the WebSocket instance
 
   useEffect(() => {
-    const ws = new WebSocket("ws://localhost:8080/ws"); // Replace with your WebSocket server URL
+    const newWs = new WebSocket("ws://localhost:8080/ws"); // Replace with your WebSocket server URL
 
-    ws.onopen = () => {
+    newWs.onopen = () => {
       console.log("WebSocket connection opened");
+      setWs(newWs); // Store the WebSocket instance
     };
 
-    ws.onmessage = (event) => {
+    newWs.onmessage = (event) => {
       const message = JSON.parse(event.data);
-      setData((prevData) => [...prevData, message]);
+
+      // Check if the message is historical data or a live data point
+       if (message.data && Array.isArray(message.data)) { 
+        // Historical data – overwrite existing data
+        setData(message.data.map((item: { timestamp: Date; value: number; }) => ({
+          timestamp: new Date(item.timestamp),
+          temp: Number(item.value) // Map "value" to "temp"
+        })));
+      } else if (message.temp) { // Check for live data structure
+        setData(prevData => [...prevData, { timestamp: new Date(Date.now()), temp: message.temp }]); // Add timestamp to live data
+    }
```

Comment:
Do you think there should be a "type" field on the JSON response that indicates whether we are getting historical or live data?

Reply:
I added a historical field (type is a keyword in go). This would introduce type safety issues to the live data packets since we martial from protobuf directly, so I just added it to historical ones and check for that